[WebAuthn] Full exclude list sent after batching during makeCredential

pascoej opened the pull request:
#### 3e8718b56658175b35cceb569d7e73560f85d0ca
<pre>
[WebAuthn] Full exclude list sent after batching during makeCredential
<a href="https://bugs.webkit.org/show_bug.cgi?id=298423">https://bugs.webkit.org/show_bug.cgi?id=298423</a>
<a href="https://rdar.apple.com/157668547">rdar://157668547</a>

Reviewed by Charlie Wolfe.

We have a bug where if we did not find a matching credential, we send the full excludeList
after checking each credential in the excludeList silently. This patch fixes that by passing
an empty exclude list in the makeCredential request.

This patch also adds testing infrastructure to make sure we don&apos;t regress this
fix. There is now the ability to verify the messages that the platform sends the
authenticator, previously we could only specify what the authenticator sends the platform.

* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https.html:
* Source/WebCore/testing/MockWebAuthenticationConfiguration.h:
* Source/WebCore/testing/MockWebAuthenticationConfiguration.idl:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::AuthenticatorManager::respondReceived):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.h:
(WebKit::AuthenticatorManager::services const):
(WebKit::AuthenticatorManager::respondReceivedInternal):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorTransportService.h:
(WebKit::AuthenticatorTransportService::validateExpectedCommandsCompleted):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockAuthenticatorManager.cpp:
(WebKit::MockAuthenticatorManager::respondReceivedInternal):
(WebKit::MockAuthenticatorManager::validateHidExpectedCommands):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockAuthenticatorManager.h:
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidConnection.cpp:
(WebKit::MockHidConnection::MockHidConnection):
(WebKit::MockHidConnection::parseRequest):
(WebKit::MockHidConnection::initializeExpectedCommands):
(WebKit::MockHidConnection::validateExpectedCommand):
(WebKit::MockHidConnection::validateAllExpectedCommandsConsumed):
(WebKit::MockHidConnection::validateExpectedCommandsCompleted):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidConnection.h:
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidService.cpp:
(WebKit::MockHidService::createHidConnection const):
(WebKit::MockHidService::validateExpectedCommandsCompleted):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidService.h:
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::continueMakeCredentialAfterCheckExcludedCredentials):

Canonical link: <a href="https://commits.webkit.org/299725@main">https://commits.webkit.org/299725@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1dbd3ca50be80ea767890b53f861a3cafa95a4b4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119969 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39662 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30313 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126299 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72042 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/59992fcb-cbfa-4e7c-bbae-02d2a592cda6) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121845 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40358 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48239 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91102 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60415 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e6ec0cf8-9063-4404-ad02-0761f41ee7fb) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122921 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32233 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107570 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71658 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c7d065a7-66b1-4c93-84ef-5fccf31ec321) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31264 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25675 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69934 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101708 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25863 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129219 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46889 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35552 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/99727 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47255 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103758 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99572 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25282 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45019 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23034 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43510 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46751 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52457 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46217 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49566 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47903 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fc6bf8890895116f1fd1270899c480ab2bcccaaf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50333

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3ffe1a5075c974e82378ff7c1df725308d77254d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50333

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/637e17003f89746d0a4ff00fcb7a795e2e950155. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50333

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/594fc5d11ef9174d6cc4ad3c180719e9d96c3661. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50333

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/72de20e03bb64604d56a51252034f3046bdbd251. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50333

pascoej committed:
[WebAuthn] Full exclude list sent after batching during makeCredential
https://bugs.webkit.org/show_bug.cgi?id=298423
rdar://157668547

Reviewed by Charlie Wolfe.

We have a bug where if we did not find a matching credential, we send the full excludeList
after checking each credential in the excludeList silently. This patch fixes that by passing
an empty exclude list in the makeCredential request.

This patch also adds testing infrastructure to make sure we don't regress this
fix. There is now the ability to verify the messages that the platform sends the
authenticator, previously we could only specify what the authenticator sends the platform.

* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https.html:
* Source/WebCore/testing/MockWebAuthenticationConfiguration.h:
* Source/WebCore/testing/MockWebAuthenticationConfiguration.idl:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::AuthenticatorManager::respondReceived):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.h:
(WebKit::AuthenticatorManager::services const):
(WebKit::AuthenticatorManager::respondReceivedInternal):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorTransportService.h:
(WebKit::AuthenticatorTransportService::validateExpectedCommandsCompleted):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockAuthenticatorManager.cpp:
(WebKit::MockAuthenticatorManager::respondReceivedInternal):
(WebKit::MockAuthenticatorManager::validateHidExpectedCommands):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockAuthenticatorManager.h:
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidConnection.cpp:
(WebKit::MockHidConnection::MockHidConnection):
(WebKit::MockHidConnection::parseRequest):
(WebKit::MockHidConnection::initializeExpectedCommands):
(WebKit::MockHidConnection::validateExpectedCommand):
(WebKit::MockHidConnection::validateAllExpectedCommandsConsumed):
(WebKit::MockHidConnection::validateExpectedCommandsCompleted):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidConnection.h:
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidService.cpp:
(WebKit::MockHidService::createHidConnection const):
(WebKit::MockHidService::validateExpectedCommandsCompleted):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidService.h:
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::continueMakeCredentialAfterCheckExcludedCredentials):

Canonical link: https://commits.webkit.org/299725@main
